### ISSUE TITLE
Initial compile fails

### ISSUE THREAD

**Issue by JackDetrick** [MicrosoftDocs/windows-uwp#677]
Severity	Code	Description	Project	File	Line	Suppression State
Error	MIDL2011	[msg]unresolved type declaration [context]: Windows.UI.Xaml.Data.INotifyPropertyChange [ RuntimeClass 'Bookstore.BookSku'  ]	Bookstore	E:\C++_WinRT\Bookstore\Bookstore\BookSku.idl	

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: a4db0aab-b906-9dec-381c-ceccc1e90c45
* Version Independent ID: 493a4c50-ffa1-5e32-c757-92306c952887
* Content: [XAML controls; bind to a C++/WinRT property - UWP app developer](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/binding-property)
* Content Source: [windows-apps-src/cpp-and-winrt-apis/binding-property.md](https://github.com/MicrosoftDocs/windows-uwp/blob/docs/windows-apps-src/cpp-and-winrt-apis/binding-property.md)
* Product: **windows**
* GitHub Login: @stevewhims
* Microsoft Alias: **stwhi**

**Comment by stevewhims** [comment]
Thanks for the message, @JackDetrick: the error you're seeing is consistent with you not having the C++/WinRT VSIX installed (in the light blue Note at the top of this topic, there's a link to info about how to get the VSIX). There are three ways of getting Windows namespace types (such as the one in the error message that you quote) to resolve: the recommended one is to install the VSIX and let that take care of automatically referencing Windows namespace types for you. Another two ways (not recommended, but I mention them out of interest) are: 1) use the /reference command line switch with the MIDL compiler, and 2) add an explicit import directive for the IDL file containing the Windows type you're referencing.

For more info on installing the VSIX, and the required versions of Visual Studio to follow the recommended path, please see https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/intro-to-using-cpp-with-winrt#visual-studio-support-for-cwinrt-and-the-vsix. For info about using MIDL 3.0 from the command-line (including the /reference switch), please see https://docs.microsoft.com/en-us/uwp/midl-3/intro (and troubleshooting info, including the error message you quote, at https://docs.microsoft.com/en-us/uwp/midl-3/troubleshooting).

**Comment by JackDetrick** [comment]
I reinstalled VSIX and got the same results. Using my copy of Photoeditor, I set "Consume Windows Runtime Extension"
to YES and received the same error (see screen shot)
________________________________

**Comment by stevewhims** [comment]
Hi, so are you able to clone and build/run the PhotoEditor sample with your Visual Studio configuration? If so, then I agree that your VSIX is functioning correctly. If the answer is yes, then you certainly don't need to turn on support for C++/CX for the PhotoEditor sample. When you set project property C/C++ > General > Consume Windows Runtime Extension to Yes, then you're asking Visual Studio to add C++/CX support to the project. You CAN mix C++/WinRT and C++/CX in the same project, but you'd do that while you were porting, and you'd have to jump through a few hoops. There's no need to do that here. If PhotoEditor is building and running immediately after a fresh clone, then your VS is good. For the project that's giving you the MIDL error, is it possible that that's not actually a C++/WinRT project? With the VSIX installed, the documented required versions of VS and the SDK, and a C++/WinRT project created in VS, then I can't account for the MIDL error you report. If you created a C++/CX project inadvertently, then I would expect the error.

Btw, I don't see a screenshot, I don't know whether these issues support that. However, if you could just copy-paste any error messages as text, that'd be good. I often have to copy-paste via NotePad into this little window, otherwise it seems to think I'm pasting an image. :) ~Steve

**Comment by JackDetrick** [comment]
I use the “Consume Windows Runtime Extensions” set to YES, so that my app will recognize a referenced winmd file. If the flag is set to NO the compiler gives me
“namespace does not exist” even though my app has set a reference to the winmd file.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Comment by stevewhims** [comment]
Hi, @JackDetrick. Personally, I think it's best to think of “Consume Windows Runtime Extensions” as meaning "This project uses C++/CX in some way". I would recommend not setting that option in a pure C++/WinRT project (pure in the sense that it uses C++/WinRT but it doesn't also use C++/CX). The project that the tutorial walks you through creating, and the C++/WinRT PhotoEditor sample app, are both pure C++/WinRT projects. So there's no reason to set “Consume Windows Runtime Extensions” to true for those projects. For a pure C++/WinRT project, you don't need to set “Consume Windows Runtime Extensions” to true in order to have your project recognize and parse a winmd file; you just reference the winmd as usual, and then the cppwinrt.exe tool (which is included in the build process as long as you have the doc'd OS/VS/SDK/VSIX versions) will take care of parsing that winmd and creating projected types for you to consume. I'm still interested in whether you're able to clone and build/run the PhotoEditor sample with your Visual Studio configuration. If so, then your config is good, and that'll be good to know. If it is, then I'd expect you to be able to create a new C++/WinRT BlankApp project, add the BookSku.idl file, paste in the content from the topic, and get a clean build.

**Comment by stevewhims** [comment]
@JackDetrick, in #686, Nick points out a typo I made in my code listing for MainPage.idl. It's possible that my typo is tripping you up (although, you did say that it's BookSku.idl that's causing problems, and not MainPage). In any case, I've corrected and updated the topic and I just merged it into master. I expect that change to be live within 24 hours. As I say, I don't know whether it's germane to the issue you're running into, so my asks in my previous reply still stand. Many thanks for your patience! ~Steve
